tests/integration/psync2.tcl

```diff
@@ -123,6 +129,25 @@ start_server {} {
             }
         }
 
+        # Waiting for replicas to inherit their new replication ID
+        for {set j 0} {$j < 5} {incr j} {
+            set role "[status $R([expr {$j}]) role]"
+            if {$role == "master"} continue
+
+            set replica_pair "$R_host($j) $R_port($j)"
+            set replica_master_host "[status $R([expr {$j}]) master_host]"
+            set replica_master_port "[status $R([expr {$j}]) master_port]"
+            set replica_master_pair "$replica_master_host $replica_master_port"
+
+            # Each replica needs to inherit the new replication ID of its master
+            wait_for_condition 100 100 {
+               [status $R($j) master_replid] == [status $R($R_id($replica_master_pair)) master_replid]
+            } else {
+                show_cluster_status
+                fail "Replica does not inherit the new replid."
+            }
+        }
+
```

Comment:
i think this is slightly more complicated than it has to be
```suggestion
        # Waiting for replicas to inherit their new replication ID
        set master_replid [status $R($master_id) master_replid]
        wait_for_condition 50 1000 {
            [status $R([expr {($master_id+1)%5}]) master_replid] == $master_replid &&
            [status $R([expr {($master_id+2)%5}]) master_replid] == $master_replid &&
            [status $R([expr {($master_id+3)%5}]) master_replid] == $master_replid &&
            [status $R([expr {($master_id+4)%5}]) master_replid] == $master_replid
        } else {
            show_cluster_status
            fail "Replicas didn't inherit master_replid"
        }

```
apart from being shorter, and not needing the `R_id` array, it also uses single timeout for the entire thing, rather than a separate timeout for each replica.

Reply:
Yes.. I wrote like this before. But there will be few problems:
- in CYCLE 1, we do not have any replication related thing  (we can add a if judge)
- in other cycle, not all instance need it.  I forgot the scene.. (the code will not pass the test)

Your code is actually my first version...

Reply:
i don't follow you.. maybe you need to explain yourself better..
eventually, all replicas will have the same ID, so a simple / single wait_for_condition can be enough (even on CYCLE1).
maybe we just don't need to pre-fetch the master's ID, and get this into the loop too. 
i.e.
```tcl
        wait_for_condition 50 1000 {
            [status $R([expr {($master_id+1)%5}]) master_replid] == [status $R($master_id) master_replid] &&
            [status $R([expr {($master_id+2)%5}]) master_replid] == [status $R($master_id) master_replid] &&
            [status $R([expr {($master_id+3)%5}]) master_replid] == [status $R($master_id) master_replid] &&
            [status $R([expr {($master_id+4)%5}]) master_replid] == [status $R($master_id) master_replid]
        } ...
```

Reply:
We actually built the replication cluster in the next test case

So in cycle 1, we actually only have 5 single instance
Means 5 completely different IDs

And in other cycle, before the slaveof no one
All replicas have the same ID.

After the slaveof no one.
And before the `# 2) Attach all the slaves to a random instance` test case

For example:
If one of the `slave` execute `slaveof no one`, became a new master
And it will have the new ID, at this time, the id is different

I didn't describe it very well.... (bad english)
Hope i made it clear (Although this code is a bit ugly)

Reply:
ohh, i see what you mean..  skipping the first cycle is easy, but indeed, before changing the topology, we wanna just wait for the ones still (directly **or indirectly**) linked to the new master to get an updated ID, not all replicas.
but does your code handles non-direct link? i think it doesn't, right?

i gave it some thought and also consulted @guybe7, our conclusion was that the best approach would be this:

there's a moment in which we have 2 masters at the same time, some replicas are still connected (directly or indirectly) to the old master, and some are already connected to the new master.
we can write some block of code that collects a list of replicas that are part of the second group (or have a boolean per replica), we do that by a recursion or a nested loop that, and then we can wait for all replicas in that second group to have the new replication id.

obviously documenting the problem (i.e. the race) and steps of the solution (i.e. the fact there are two masters, and two groups possibly with indirect replication).

sound right?

Reply:
cool. Better one.  i will try this way later.

Reply:
If I understand it correctly, we need five list. (or a map)
eg list1, list2, ... list5.

Store the possible slaves of the each instance.

So that after we call `slaveof no one`
we can track all the slaves of the new master
and do a loop to make sure every slave get the new ID

like (python code):
```python
    def ensure_self_slave_get_new_id(master_id):
        slaves = [xx, xx, xx]  # we get it from a map

        for slave in slaves:
            wait for the slave get the new ID

            ensure_self_slave_get_new_id(slave.id)


    master_id = 1
    ensure_self_slave_get_new_id(master_id)
```

I understand right?

Reply:
there are many ways to write it.
one may be similar to what you wrote, but i'd rather put the wait_for_condition outside, and have it call a method that returns true or false.

one (that doesn't declare new functions) can be:
```
# build a lookup with the root master of each replica (head of the chain)
root_master = []
for replica in replicas:
   r = replica
   while true:
      if r.masterport == 0
            root_master[replica.id] = r.id
            break
      master_id = replica_id_from_port[r.masterport] ;# you'll need to prepare that lookup ahead of time too
      r = replicas[master_id]

# for the newly detached master-replica chain (new master and existing replicas that were
# already connected to it, to get updated on the new replication id
wait_for_condition 50 1000 {
    ([status $R(0]) master_replid] == [status $R([root_master[0]]) master_replid]) &&
    ([status $R(1]) master_replid] == [status $R([root_master[1]]) master_replid]) &&
    ([status $R(2]) master_replid] == [status $R([root_master[2]]) master_replid]) &&
    ([status $R(3]) master_replid] == [status $R([root_master[3]]) master_replid]) &&
    ([status $R(4]) master_replid] == [status $R([root_master[4]]) master_replid])
} ...
````

Reply:
I actually spent some time understanding your way. 
Have to say it is very clever... Thanks a lot!

I have translated it into code. Hope i do it right